Correct parameter index for Polygonal in DataSurfaces

rraustad opened the pull request:
## Pull request overview

Addresses #5507 
Fixes polygons being classified as triangular windows for some daylighting calculations.
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [x] Code style (parentheses padding, variable names)
- [x] Functional code review (it has to work!)
- [x] If defect, results of running current develop vs this branch should exhibit the fix
- [x] CI status: all green or justified
- [x] Performance: CI Linux results include performance check -- verify this
- [x] Unit Test(s)
- C++ checks:
  - [x] Argument types
  - [x] If any virtual classes, ensure virtual destructor included, other things
- IDD changes N/A

rraustad committed:
Correct parameter index for Polygonal in DataSurfaces

rraustad committed:
converted to enum and continued with unit test changes

rraustad commented:
Converted to enum. Need someone who knows the syntax to review.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5507-Polygonal-and-TriangularDoor-have-the-same-value

rraustad committed:
Included all surface shape types in unit test

mjwitte commented:
@Myoldmopar @mbadams5 Does this pass syntax/style review?

mbadams5 committed:
Convert enum to enum class

mbadams5 commented:
@mjwitte @rraustad I just pushed up a commit to convert the `enum` to an `enum class`. The `enum` usage was correct, but we should prefer `enum class` whenever possible. It will be a little bit more code changes than just converting the integers to an `enum` as seen in my commit. However, `enum class` has many advantages over a plain `enum`, namely: type safety, no name collisions, no implicit conversion to int, can choose different underlying type (int, char, unsigned int, etc).

rraustad commented:
"It will be a little bit more code changes" is an understatement. It is good to see how to do this correctly. This mostly makes sense. Thanks Mark.

Can we grab these changes and put them in a wiki somewhere?

mjwitte commented on tst/EnergyPlus/unit/DataSurfaces.unit.cc:
```diff
@@ -84,6 +84,349 @@ using namespace EnergyPlus::SurfaceGeometry;
 using namespace ObjexxFCL;
 using DataVectorTypes::Vector;
 
+TEST_F( EnergyPlusFixture, DataSurfaces_ProcessSurfaceVertices )
```
@rraustad Since ProcessSurfaceVertices is in SurfaceGeometry, this unit test should be there as well, next to the current one which exercises ProcessSurfaceVertices in a different way.  Good to have both forms.

rraustad committed:
Corrected issue with overhangs and fins. Does any example file fail now?

rraustad commented:
I noticed that the code produces a warning when the right (left) inputs are 0 for Shading:Fin:Projection.

```
 Shading:Fin:Projection,",
    Surface 6 - RectangularLeftFin,    !- Name",
    Surface 3 - Rectangle,   !- Window or Door Name",
    .01,                     !- Left Extension from Window/Door {m}",
    .01,                     !- Left Distance Above Top of Window {m}",
    .01,                     !- Left Distance Below Bottom of Window {m}",
    90,                      !- Left Tilt Angle from Window/Door {deg}",
    .1,                      !- Left Depth as Fraction of Window/Door Width {dimensionless}",
    0,                       !- Right Extension from Window/Door {m}",
    0,                       !- Right Distance Above Top of Window {m}",
    0,                       !- Right Distance Below Bottom of Window {m}",
    0,                       !- Right Tilt Angle from Window/Door {deg}",
    0;                       !- Right Depth as Fraction of Window/Door Width {dimensionless}",
```

Warning in 2 places (5203 and 5290 of SurfaceGeometry):

```
MakeFin = true;
if ( Length * Depth <= 0.0 ) {
    ShowWarningError( cCurrentModuleObject + "=Right Fin of \"" + cAlphaArgs( 1 ) + "\", illegal surface area=[" + RoundSigDigits( Length * Depth, 2 ) + "]. Surface will NOT be entered." );
    MakeFin = false;
}
```

Add check to see if other fin has non-zero area before issuing warning ???

mjwitte commented:
No, don't need to check for zero for the other inputs - they all are valid as zero.

rraustad commented:
Yes, those entries are valid as 0. The question is if the warning is necessary when one fin is entered > 0 and the other is 0. Right now a warning is issued for the 0 entered fin. So in the above example you get a warning for the right fin since the area = 0. In my unit test, I also see a warning for Surface 7 - RectangularRightFin since the left inputs are 0.

mjwitte commented:
I think it's fine as-is.

rraustad commented:
The first failure here shows a change in azimuth, which I think I corrected. The left and right fin both have an azimuth of 270. I would think the left and right fin would be 180 degrees different. The first eio line shows the left fin at 270 with the mirror of this surface at 90. This makes sense. It then shows the same for the right fin, which I thought was incorrect. I corrected this and now the right fin has 90 and 270 for the fin and mirrored fin, respectively. 

```
Shading_Surface,ZN001:WALL001:SHADE003 Left,Shading,ZN001:WALL001,None,,0.0,0.0, ,15.60,15.60,15.60,270.00,90.00,3.00,5.20,,,,,,,,,,,4,9.90,6.00,8.10,9.90,6.00,2.90,9.90,3.00,2.90,9.90,3.00,8.10
Shading_Surface,Mir-ZN001:WALL001:SHADE003 Left,Shading,ZN001:WALL001,None,,0.0,0.0, ,15.60,15.60,15.60,90.00,90.00,3.00,5.20,,,,,,,,,,,4,9.90,3.00,8.10,9.90,3.00,2.90,9.90,6.00,2.90,9.90,6.00,8.10
-Shading_Surface,ZN001:WALL001:SHADE003 Right,Shading,ZN001:WALL001,None,,0.0,0.0, ,15.60,15.60,15.60,270.00,90.00,3.00,5.20,,,,,,,,,,,4,13.10,6.00,8.10,13.10,6.00,2.90,13.10,3.00,2.90,13.10,3.00,8.10
-Shading_Surface,Mir-ZN001:WALL001:SHADE003 Right,Shading,ZN001:WALL001,None,,0.0,0.0, ,15.60,15.60,15.60,90.00,90.00,3.00,5.20,,,,,,,,,,,4,13.10,3.00,8.10,13.10,3.00,2.90,13.10,6.00,2.90,13.10,6.00,8.10
+Shading_Surface,ZN001:WALL001:SHADE003 Right,Shading,ZN001:WALL001,None,,0.0,0.0, ,15.60,15.60,15.60,90.00,90.00,3.00,5.20,,,,,,,,,,,4,13.10,6.00,8.10,13.10,6.00,2.90,13.10,9.00,2.90,13.10,9.00,8.10
+Shading_Surface,Mir-ZN001:WALL001:SHADE003 Right,Shading,ZN001:WALL001,None,,0.0,0.0, ,15.60,15.60,15.60,270.00,90.00,3.00,5.20,,,,,,,,,,,4,13.10,9.00,8.10,13.10,9.00,2.90,13.10,6.00,2.90,13.10,6.00,8.10
```

rraustad commented:
Here's a look at the same data from V8.4 eio for 4ZoneWithShading_Simple_1.

![4zonewithshading_simple_1](https://cloud.githubusercontent.com/assets/5778156/13752883/2250f016-e9e7-11e5-906e-2f5bed5f27a6.png)

mjwitte commented:
I agree that the left and right fins should have mirrored azimuths (relative to the window), but this should only matter if shading surface mirroring is turned off.  I did not expect any big eso diffs out of that change, because the shading should come out the same as before.

rraustad commented:
So if the fin and mirrored fin have azimuths 180 degrees out, and the right fin also has these same properties, then it doesn't matter if each fin has the same azimuth since the mirror will create a surface that is 180 degrees different from the actual fin. Then I wonder why there were differences with this subtle change??

mjwitte commented:
Exactly.  Only perpedicular fins will differ by 180.  Angled fins will differ by some other value.

rraustad commented:
The only other change for this check in was to correct negative angles, in 2 places.

```
if ( SurfaceTmp( SurfNum ).Azimuth < 0.0 ) SurfaceTmp( SurfNum ).Azimuth += 360.0;
```

So one (fix azimuth) or the other (negative angles) change caused these diff's.

I'll back out the negative angle change and try again?

rraustad commented:
If I comment out the change at line 1270, where I set the base surface number for overhangs and fins, the diff's are much less. They are not 0, I guess since I changed the math on the azimuth angle.

line 1270 of SurfaceGeometry:

```
for ( SurfNum = 1; SurfNum <= TotSurfaces; ++SurfNum ) {  
  if ( Surface( SurfNum ).Class != SurfaceClass_Shading ) continue;  
  Surface( SurfNum ).BaseSurf = FindItemInList( Surface( SurfNum ).BaseSurfName, Surface, TotSurfaces );  
}  
```

Of course this will break the unit test. The question is if fins and overhangs should have a base surface index or not. They do hang off the building so what would the base surface properties be used for?

rraustad commented:
Reverting the math on the fin azimuth angle gets back to 0 diff's. Obvious answer since this is where I started.

```
4ZoneWithShading_Simple_1.csv
 vs
4ZoneWithShading_Simple_1_new.csv

All Records are Equal.    48 records compared
```

So I can check this in like this or ?????

mjwitte commented:
How about checking in with azimuth changes, but without the base surface search.  And merge in develop to current.

rraustad commented:
There will still be big diff's, just not as big as before. I saw 7% on zone cooling rate vs a 26% number before (although I can't recall which variable that was).

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5507-Polygonal-and-TriangularDoor-have-the-same-value

rraustad committed:
Reverted setting of BaseSurf for overhangs and fins. Change to azimuth angle for fins is still used in this version. Unit Tests commented to show failures (and uses EXPECT_NE).

mjwitte commented:
Then the best path is to separate these issues.  Let's get the SurfaceShape enum in without any changes to the fin code.  You can post a followup issue to look at after v8.5.0 release.

rraustad commented:
There are only 2 files with diff's. These are both the 4ZoneWithShading_Simple_*.

It's the shading surfaces that are different.
Here's one example with V8.4 shown first:

```
==================================
Surface=ZN001:WALL001:SHADE003 Right is used as Receiving Surface in calculations and is convex.
Number of general casting surfaces=4
..Surface=ZN001:WALL001:WIN001:SHADE001
..Surface=Mir-ZN001:WALL001:SHADE003 Left
..Surface=ZN001:WALL001
..Surface=ZN003:WALL001
Number of back surfaces=0
Number of receiving sub surfaces=0
==================================

==================================
Surface=Mir-ZN001:WALL001:SHADE003 Right is used as Receiving Surface in calculations and is convex.
Number of general casting surfaces=2
..Surface=ZN001:WALL001:WIN001:SHADE001
..Surface=Mir-ZN001:WALL001:SHADE003 Left
Number of back surfaces=0
Number of receiving sub surfaces=0
==================================
```

rraustad committed:
Revert change in fin azimuth angle to yield identical results for enum SurfaceShape issue.

mjwitte commented:
I can't explain why, but these azimuths are rotating the fin about and edge, so things aren't quite what they seem at this point in the code.  4ZoneWithShading_Simple_1.dxf output before
![image](https://cloud.githubusercontent.com/assets/4991075/13765218/233ae96c-ea25-11e5-9728-cb09765a1f56.png)

and after
![image](https://cloud.githubusercontent.com/assets/4991075/13765221/30701648-ea25-11e5-9e27-2c7ab751a9bb.png)

So, the azimuth changes need to come back out.

rraustad commented:
@Myoldmopar @mjwitte This is ready.

mjwitte commented:
@rraustad Sorry, but I'd rather see the fin section of SurfaceGeometry reverted back entirely - remove the commented blocks and remove the azimuth+360 lines.  I know they seem innocuous, but azimuth isn't what it seems to be here.

rraustad committed:
Remove commented lines and correction of azimuth to positive values.

rraustad commented:
Your probably right about getting rid of azimuth+360. That might have flipped the fins inward on the North side of the building.

rraustad commented:
I used Output:Diagnostics, DoNotMirrorAttachedShading (not in IDD) and got different results.

![donotmirrorattachedshading](https://cloud.githubusercontent.com/assets/5778156/13784487/d3acd528-eaa5-11e5-948c-e8d048c655ad.png)

mjwitte commented:
@rraustad Let's get the polygon fix in place and please start a new issue for the fin geometry and copy some of this discussion over there.

rraustad commented:
Follow up issue #5550
